src/utils/i18n.ts

```diff
@@ -0,0 +1,264 @@
+  for (let i = 0; i < normalizedPreferredlocales.length; i++) {
+    const current = normalizedPreferredlocales[i];
+
+    possibleLocales.push(current);
+
+    // Create subsets of current and add them to the list of possible locales.
+    // Only add them if they aren't subsets of the next locale or if they appear later in the
+    // preference list.
+    const next = normalizedPreferredlocales[i + 1] || '';
+    const subsets = makeSubsets(current).filter((s) => !baseSet[s] && next.search(s) === -1);
+
+    possibleLocales.push(...subsets);
+  }
+
+  const matchedLocales =
+    possibleLocales
+      // Swap any locales that have aliases with their alias.
+      .map((l) => {
+        const alias = normalizedAliases[l];
+        return typeof alias === 'string' ? alias : l;
+      })
+      // Filter for a locale that is in the supported set.
+      .filter((l) => supportedSet[l] === true);
+
+  // Return the first locale that matched (or undefined if it is not available).
+  return matchedLocales[0];
+}
+
+function isAlias(aliases: Aliases | undefined): aliases is Aliases {
+  return typeof aliases === 'object';
+}
+
+function isTranslations(translations: Translations[string]): translations is Translations {
+  return typeof translations === 'object';
+}
+
+/**
+ * Generates viable subsets from a given locale.
+ *
+ * Examples:
+ * ```
+ * new LanguageTag('zh-Hant-HK').makeSubsets()    // ['zh-Hant', 'zh']
+ * new LanguageTag('zh-HK').makeSubsets()         // ['zh']
+ * new LanguageTag('zh-Hant').makeSubsets()       // ['zh']
+ * new LanguageTag('zh').makeSubsets()            // []
+ * ```
+ */
+function makeSubsets(locale: string): string[] {
+  const subsets: string[] = [];
+  const languageTag = new LanguageTag(locale);
+
+  if (languageTag.hasRegion() && languageTag.hasScript()) {
+    subsets.push(`${languageTag.language}-${languageTag.script}`);
+  }
+
+  if (languageTag.language !== languageTag.toString()) {
+    subsets.push(languageTag.language);
+  }
+
+  return subsets;
+}
+
+/**
+ * Canonicalize a single locale.
+ */
+export function getCanonicalLocale(locale: string): string {
+  return LanguageTag.toString(locale);
+}
+
+/**
+ * Returns an array of canonicalized locales. This method does what the specification for
+ * Intl.getCanonicalLocales does.
+ *
+ * ```
+ * getCanonicalLocales('EN-US'); // ["en-US"]
+ * getCanonicalLocales(['EN-US', 'Fr']); // ["en-US", "fr"]
+ * ```
+ */
+export function getCanonicalLocales(locales: string | string[]): string[] {
+  if (typeof locales === 'string') {
+    return [getCanonicalLocale(locales)];
+  } else {
+    return locales.map(getCanonicalLocale);
+  }
+}
+
+/**
+ * Look up a translation with the given key and locales.
+ *
+ * NOTE: Locales that are passed into this method are not normalized. It is important do this if
+ * the list of locales are arbitrary (for example, an end users browser). Look into using
+ * `getCanonicalLocales` on locales, rule keys and translation keys.
```

Comment:
Footnote saying to canonicalize the locales. As mentioned in the description, it seems inefficient to do this everytime `t` is called.